File: crates/nmp-cli/registry/swiftui/content-view/NostrContentView.swift
import AVKit
import SwiftUI
import UIKit

/// SwiftUI renderer for a `ContentTreeWire`. Walks `tree.roots`, flattens the
/// arena into block-level groups via `nostrContentGroups`, and renders each
/// block (paragraph/heading/media/code/list/quote/rule/image/event-ref/
/// placeholder) with SwiftUI `Text` concatenation for inline runs.
///
/// Data injection contract:
///   • Theming + tap callbacks come from `NostrContentRenderer` in the
///     SwiftUI environment (see `swiftui/content-core`).
///   • Mention display labels are provided by the app via `mentionLabel`.
///   • Embedded events (`nostr:nevent…` / `nostr:naddr…`) render through the
///     kind-dispatch registry (ADR-0034): the app binds an `EmbedEnvelopeSource`
///     + claim sink + `NostrKindRegistry` via `.embedEnvelopeSource(...)` and
///     `eventRefView` dispatches through `EmbeddedEvent`. With no host bound,
///     `EmbeddedEvent` still renders a loading placeholder via the default
///     registry — there is no separate quote-card fallback.
public struct NostrContentView: View {
    public var tree: ContentTreeWire
    public var font: Font
    public var mentionLabel: (NostrWireUri) -> String
    /// Highlight overlays to paint on the body. Each decoration's `quote` is
    /// matched against the rendered plain text and painted with its colour;
    /// taps route to `renderer.callbacks.onDecorationTap`. Empty (the default)
    /// keeps the fast `Text`-concatenation path.
    public var decorations: [NostrContentDecoration]
    /// Opt-in body text selection. When `true`, paragraph/heading runs become
    /// selectable and the selection edit menu gains a "Highlight" action that
    /// fires `renderer.callbacks.onTextSelected(quote, context)`. Off by
    /// default so non-article surfaces keep the lightweight `Text` path.
    public var selectionEnabled: Bool

    // Non-private so the article-mode extension (NostrContentArticleView,
    // NostrContentAttributed) in the same module can read theming + callbacks.
    @Environment(\.nostrContentRenderer) var renderer
    @Environment(\.embedEnvelopeSource) private var embedEnvelopeSource
    @Environment(\.embedClaimSink) private var embedClaimSink
    @Environment(\.nostrKindRegistry) private var nostrKindRegistry

    public init(
        tree: ContentTreeWire,
        font: Font = .body,
        mentionLabel: @escaping (NostrWireUri) -> String = NostrContentView.defaultMentionLabel,
        decorations: [NostrContentDecoration] = [],
        selectionEnabled: Bool = false
    ) {
        self.tree = tree
        self.font = font
        self.mentionLabel = mentionLabel
        self.decorations = decorations
        self.selectionEnabled = selectionEnabled
    }

    public var body: some View {
        let groups = nostrContentGroups(tree)
        if groups.isEmpty {
            EmptyView()
        } else if articleMode {
            // Footnote-definition paragraphs (`[^1]: …`) are rendered once, by
            // `footnoteSection`. Suppress the originals from the normal loop so
            // the body shows each definition exactly once and the marker
            // scrolls to the single rendered block.
            let bodyGroups = groups.filter { !isFootnoteDefinitionGroup($0) }
            ScrollViewReader { proxy in
                VStack(alignment: .leading, spacing: 8) {
                    ForEach(Array(bodyGroups.enumerated()), id: \.offset) { _, group in
                        groupView(group, proxy: proxy)
                    }
                    footnoteSection(proxy: proxy)
                }
            }
        } else {
            VStack(alignment: .leading, spacing: 8) {
                ForEach(Array(groups.enumerated()), id: \.offset) { _, group in
                    groupView(group, proxy: nil)
                }
            }
        }
    }

    // MARK: - Group dispatch

    @ViewBuilder
    private func groupView(_ group: NostrContentGroup, proxy: ScrollViewProxy?) -> some View {
        switch group {
        case .inline(let level, let children):
            inlineGroup(level: level, children: children, proxy: proxy)
        case .media(let urls, let kind):
            mediaGroup(urls: urls, kind: kind)
        case .eventRef(let uri):
            eventRefView(uri)
        case .codeBlock(let info, let body):
            codeBlockView(info: info, body: body)
        case .blockQuote(let children):
            blockQuoteView(children: children)
        case .list(let orderedStart, let items):
            listView(orderedStart: orderedStart, items: items)
        case .rule:
            ruleView
        case .image(let alt, let title, let src):
            imageBlockView(alt: alt, title: title, src: src)
        case .placeholder(let reason):
            placeholderChip(reason: reason)
        }
    }

    @ViewBuilder
    private func inlineGroup(level: NostrContentInlineLevel, children: [UInt32], proxy: ScrollViewProxy?) -> some View {
        if articleMode {
            articleInlineGroup(level: level, children: children, proxy: proxy)
        } else {
            let concatenated = children.reduce(Text("")) { acc, child in
                acc + inlineText(child)
            }
            switch level {
            case .paragraph:
                concatenated
                    .font(font)
                    .foregroundStyle(renderer.textColor)
                    .fixedSize(horizontal: false, vertical: true)
            case .heading(let lvl):
                concatenated
                    .font(headingFont(for: lvl))
                    .foregroundStyle(renderer.textColor)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
    }

    // MARK: - Inline text concatenation

    /// Convert one arena node into a `Text` value that can be `+`-concatenated
    /// with neighbouring runs. Recursive children (emphasis/strong/link/heading
    /// /paragraph) are walked here so the whole inline subtree collapses to a
    /// single `Text`.
    public func inlineText(_ index: UInt32) -> Text {
        if index == nostrContentNewlineSentinel { return Text("\n") }
        guard let node = tree.node(at: index) else { return Text("") }
        switch node {
        case .text(let value):
            return Text(value)
        case .mention(let uri):
            return Text("@\(mentionLabel(uri))")
                .foregroundStyle(renderer.mentionColor)
                .bold()
        case .eventRef(let uri):
            return Text("↩ \(shortEntity(uri.primaryId))")
                .foregroundStyle(renderer.linkColor)
                .bold()
        case .hashtag(let tag):
            return Text("#\(tag)")
                .foregroundStyle(renderer.hashtagColor)
                .bold()
        case .url(let value):
            return Text(value).foregroundStyle(renderer.linkColor)
        case .emoji(let shortcode, _):
            // Apps fill `renderer.emojiImages` from kind:0 / NIP-30 tag data
            // and the inline run renders the image directly. Empty dict
            // (the default) falls back to the literal `:shortcode:` text so
            // unwired apps still get a readable surface.
            if let img = renderer.emojiImages[shortcode] {
                return Text(Image(uiImage: img))
            }
            return Text(":\(shortcode):")
        case .invoice:
            return Text("⚡ invoice").foregroundStyle(renderer.linkColor)
        case .emphasis(let children):
            return children.reduce(Text("")) { $0 + inlineText($1).italic() }
        case .strong(let children):
            return children.reduce(Text("")) { $0 + inlineText($1).bold() }
        case .inlineCode(let value):
            return Text(value).font(.system(.body, design: .monospaced))
        case .link(let children, let href):
            let label = children.reduce(Text("")) { $0 + inlineText($1) }
            if let href, !href.isEmpty {
                return label
                    .foregroundStyle(renderer.linkColor)
                    .underline()
            }
            return label
        case .image(let alt, _, _):
            return Text(alt.isEmpty ? "[image]" : "[\(alt)]")
                .foregroundStyle(renderer.placeholderColor)
        case .softBreak:
            return Text(" ")
        case .hardBreak:
            return Text("\n")
        case .paragraph(let children),
             .heading(_, let children),
             .blockQuote(let children):
            return children.reduce(Text("")) { $0 + inlineText($1) }
        case .list, .codeBlock, .rule, .media, .placeholder:
            // Block-level — should never appear inside an inline reduce. Emit
            // nothing rather than break the text concatenation.
            return Text("")
        }
    }

    // MARK: - Block builders

    @ViewBuilder
    private func mediaGroup(urls: [String], kind: NostrMediaKind) -> some View {
        switch kind {
        case .image:
            let parsed = urls.compactMap(URL.init(string:))
            if !parsed.isEmpty {
                NostrMediaGrid(imageUrls: parsed)
            }
        case .video:
            // Inline playback via `AVKit.VideoPlayer` so video media nodes
            // render with native scrub / fullscreen controls. Audio stays on
            // the compact link-style row (no waveform UI in v1).
            if let first = urls.first.flatMap(URL.init(string:)) {
                VideoPlayer(player: AVPlayer(url: first))
                    .aspectRatio(16.0 / 9.0, contentMode: .fit)
                    .clipShape(RoundedRectangle(cornerRadius: 10))
            }
        case .audio:
            if let first = urls.first.flatMap(URL.init(string:)) {
                mediaRow(first, systemImage: "speaker.wave.2.fill")
            }
        }
    }

    private func mediaRow(_ url: URL, systemImage: String) -> some View {
        Button {
            renderer.callbacks.onLinkTap(url)
        } label: {
            HStack(spacing: 10) {
                Image(systemName: systemImage)
                    .font(.title2)
                    .foregroundStyle(renderer.linkColor)
                Text(url.lastPathComponent)
                    .font(.caption.monospaced())
                    .foregroundStyle(renderer.secondaryTextColor)
                    .lineLimit(1)
                Spacer(minLength: 0)
            }
            .padding(12)
            .frame(maxWidth: .infinity)
            .background(renderer.codeBackgroundColor, in: RoundedRectangle(cornerRadius: 8))
        }
        .buttonStyle(.plain)
    }

    private func eventRefView(_ uri: NostrWireUri) -> some View {
        // Kind-dispatch path (ADR-0034 / F-CR-04). `EmbeddedEvent` owns the
        // claim/release lifecycle via `task(id:)` + `onDisappear`; the registry
        // picks the renderer for the resolved projection. The view always
        // renders even if the host hasn't resolved the envelope yet (loading
        // placeholder via `EmbedChromeContainer`). With no registry bound, the
        // built-in defaults render.
        EmbeddedEvent(
            uri: uri.uri,
            envelope: embedEnvelopeSource?.envelopeForPrimaryID(uri.primaryId)
                ?? embedEnvelopeSource?.envelopeForURI(uri.uri),
            registry: nostrKindRegistry ?? NostrKindRegistry.makeDefault(),
            claimSink: embedClaimSink
        )
    }

    private func codeBlockView(info: String?, body: String) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            if let info, !info.isEmpty {
                Text(info)
                    .font(.caption2.monospaced())
                    .foregroundStyle(renderer.secondaryTextColor)
            }
            Text(body)
                .font(.system(.callout, design: .monospaced))
                .foregroundStyle(renderer.textColor)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(10)
        .background(renderer.codeBackgroundColor, in: RoundedRectangle(cornerRadius: 6))
    }

    private func blockQuoteView(children: [UInt32]) -> some View {
        let text = children.reduce(Text("")) { $0 + inlineText($1) }
        return HStack(spacing: 10) {
            Rectangle()
                .fill(renderer.quoteBorderColor)
                .frame(width: 3)
            text
                .font(font.italic())
                .foregroundStyle(renderer.secondaryTextColor)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, 4)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    @ViewBuilder
    private func listView(orderedStart: UInt64?, items: [[UInt32]]) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            ForEach(Array(items.enumerated()), id: \.offset) { offset, children in
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text(marker(orderedStart: orderedStart, offset: offset))
                        .font(font)
                        .foregroundStyle(renderer.secondaryTextColor)
                    children.reduce(Text("")) { $0 + inlineText($1) }
                        .font(font)
                        .foregroundStyle(renderer.textColor)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
    }

    private func marker(orderedStart: UInt64?, offset: Int) -> String {
        if let orderedStart {
            return "\(orderedStart + UInt64(offset))."
        }
        return "•"
    }

    private var ruleView: some View {
        Rectangle()
            .fill(renderer.quoteBorderColor)
            .frame(height: 1)
            .padding(.vertical, 4)
    }

    private func imageBlockView(alt: String, title: String?, src: String?) -> some View {
        if let src, let url = URL(string: src) {
            return AnyView(
                NostrMediaGrid(imageUrls: [url])
                    .accessibilityLabel(title ?? alt)
            )
        }
        return AnyView(
            Text(alt.isEmpty ? "[image]" : "[\(alt)]")
                .font(.caption)
                .foregroundStyle(renderer.placeholderColor)
        )
    }

    private func placeholderChip(reason: NostrWirePlaceholderReason) -> some View {
        let label: String
        let icon: String
        switch reason {
        case .depthLimit:
            label = "Nested content collapsed"
            icon = "chevron.down.square"
        case .unresolvedUri:
            label = "Unresolved reference"
            icon = "questionmark.square.dashed"
        }
        return HStack(spacing: 6) {
            Image(systemName: icon)
                .foregroundStyle(renderer.placeholderColor)
            Text(label)
                .font(.caption.weight(.semibold))
                .foregroundStyle(renderer.placeholderColor)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(renderer.codeBackgroundColor, in: Capsule())
    }

    // MARK: - Defaults / helpers

    public nonisolated static func defaultMentionLabel(_ uri: NostrWireUri) -> String {
        let value = uri.primaryId
        guard value.count > 12 else { return value }
        return "\(value.prefix(8))…\(value.suffix(4))"
    }

    private func headingFont(for level: UInt8) -> Font {
        switch level {
        case 1: return .largeTitle.weight(.bold)
        case 2: return .title.weight(.bold)
        case 3: return .title2.weight(.semibold)
        case 4: return .title3.weight(.semibold)
        case 5: return .headline
        default: return .subheadline.weight(.semibold)
        }
    }

    private func shortEntity(_ value: String) -> String {
        guard value.count > 12 else { return value }
        return "\(value.prefix(8))…\(value.suffix(4))"
    }
}

File: apps/nmp-gallery/android/app/src/main/kotlin/org/nmp/gallery/registry/NostrContentView.kt
// Requires: compose-ui, compose-foundation, compose-material3,
// androidx.compose.material:material-icons-extended (for ExpandMore,
// HelpOutline, PlayArrow, VolumeUp). Pulls `compose/content-media-grid` and
// `compose/content-kind-registry` as registry deps. Kotlin 1.9+.
//
// Compose mirror of the SwiftUI `NostrContentView`. Walks a
// `ContentTreeWire`, flattens the arena into block-level groups via
// `nostrContentGroups`, and renders each block (paragraph / heading / media /
// code / list / quote / rule / image / event-ref / placeholder).
//
// Data injection contract:
//   - Theming + tap callbacks come from `LocalNostrContentRenderer`
//     (see `compose/content-core`).
//   - Mention display labels are provided by the app via `mentionLabel`.
//   - Embedded events (`nostr:nevent…` / `nostr:naddr…`) render via the
//     kind-dispatch registry (ADR-0072): `EventRefBlock` delegates to the
//     `EmbeddedEvent` composable from `compose/content-kind-registry`, which
//     resolves the URI, reads the kernel-resolved `EmbeddedEventEnvelope` from
//     `LocalResolvedEventEmbeds`, and dispatches to the per-kind renderer in
//     `LocalNostrKindRegistry`.
//
// Inline runs are flattened into a single `AnnotatedString` (with per-run
// styling) and shown as a `ClickableText` so tap-offset → annotation routing
// can dispatch the matching callback. Block nodes use `Column`.

package org.nmp.gallery.registry

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.text.ClickableText
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

@Composable
public fun NostrContentView(
    tree: ContentTreeWire,
    modifier: Modifier = Modifier,
    textStyle: TextStyle = MaterialTheme.typography.bodyLarge,
    mentionLabel: (WireNostrUri) -> String = ::defaultMentionLabel,
) {
    val groups = nostrContentGroups(tree)
    if (groups.isEmpty()) return

    Column(
        verticalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier,
    ) {
        for (group in groups) {
            RenderGroup(
                group = group,
                tree = tree,
                textStyle = textStyle,
                mentionLabel = mentionLabel,
            )
        }
    }
}

public fun defaultMentionLabel(uri: WireNostrUri): String {
    val value = uri.primaryId
    if (value.length <= 12) return value
    return "${value.take(8)}…${value.takeLast(4)}"
}

// ---------------------------------------------------------------------------
// Block dispatch
// ---------------------------------------------------------------------------

@Composable
private fun RenderGroup(
    group: NostrContentGroup,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
) {
    when (group) {
        is NostrContentGroup.Inline -> InlineGroup(
            level = group.level,
            children = group.children,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        is NostrContentGroup.MediaGroup -> MediaGroupBlock(
            urls = group.urls,
            kind = group.kind,
        )
        is NostrContentGroup.EventRefGroup -> EventRefBlock(
            uri = group.uri,
        )
        is NostrContentGroup.CodeBlockGroup -> CodeBlockBlock(
            info = group.info,
            body = group.body,
        )
        is NostrContentGroup.BlockQuoteGroup -> BlockQuoteBlock(
            children = group.children,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        is NostrContentGroup.ListGroup -> ListBlock(
            orderedStart = group.orderedStart,
            items = group.items,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        NostrContentGroup.RuleGroup -> RuleBlock()
        is NostrContentGroup.ImageGroup -> ImageBlock(
            alt = group.alt,
            title = group.title,
            src = group.src,
        )
        is NostrContentGroup.PlaceholderGroup -> PlaceholderChip(reason = group.reason)
    }
}

// ---------------------------------------------------------------------------
// Inline rendering (Text + Text concatenation → AnnotatedString)
// ---------------------------------------------------------------------------

@Composable
private fun InlineGroup(
    level: NostrContentInlineLevel,
    children: List<UInt>,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
) {
    val renderer = LocalNostrContentRenderer.current
    val annotated = buildAnnotatedString {
        for (child in children) {
            appendInline(
                index = child,
                tree = tree,
                renderer = renderer,
                mentionLabel = mentionLabel,
            )
        }
    }
    val effectiveStyle = when (level) {
        NostrContentInlineLevel.Paragraph -> textStyle.copy(color = renderer.textColor)
        is NostrContentInlineLevel.Heading -> headingStyle(level.level).copy(color = renderer.textColor)
    }
    ClickableText(
        text = annotated,
        style = effectiveStyle,
        modifier = Modifier.fillMaxWidth(),
        onClick = { offset ->
            dispatchInlineTap(annotated, offset, renderer)
        },
    )
}

/**
 * Append one arena node's inline projection to the [AnnotatedString] builder.
 * Recursive children (emphasis / strong / link / heading / paragraph) are
 * walked here so the whole inline subtree collapses into the running text.
 * Block-level nodes (`list`, `code_block`, `rule`, `media`, `placeholder`)
 * appearing inside an inline group emit nothing rather than break flow.
 */
internal fun AnnotatedString.Builder.appendInline(
    index: UInt,
    tree: ContentTreeWire,
    renderer: NostrContentRenderer,
    mentionLabel: (WireNostrUri) -> String,
) {
    if (index == NOSTR_CONTENT_NEWLINE_SENTINEL) {
        append('\n')
        return
    }
    val node = tree.nodeAt(index) ?: return
    when (node) {
        is WireNode.Text -> append(node.text)
        is WireNode.Mention -> {
            val label = mentionLabel(node.uri)
            withAnnotationScope(MENTION_ANNOTATION, node.uri.primaryId) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.mentionColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("@$label") }
            }
        }
        is WireNode.EventRef -> {
            val short = shortEntity(node.uri.primaryId)
            withAnnotationScope(EVENT_REF_ANNOTATION, node.uri.primaryId) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.linkColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("↩ $short") }
            }
        }
        is WireNode.Hashtag -> {
            withAnnotationScope(HASHTAG_ANNOTATION, node.tag) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.hashtagColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("#${node.tag}") }
            }
        }
        is WireNode.Url -> {
            withAnnotationScope(LINK_ANNOTATION, node.url) {
                withStyleScope(SpanStyle(color = renderer.linkColor)) {
                    append(node.url)
                }
            }
        }
        is WireNode.Emoji -> append(":${node.shortcode}:")
        is WireNode.Invoice -> {
            withStyleScope(SpanStyle(color = renderer.linkColor)) {
                append("⚡ invoice")
            }
        }
        is WireNode.Emphasis -> withStyleScope(SpanStyle(fontStyle = FontStyle.Italic)) {
            for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        }
        is WireNode.Strong -> withStyleScope(SpanStyle(fontWeight = FontWeight.Bold)) {
            for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        }
        is WireNode.InlineCode -> withStyleScope(SpanStyle(fontFamily = FontFamily.Monospace)) {
            append(node.code)
        }
        is WireNode.Link -> {
            val href = node.href
            if (!href.isNullOrEmpty()) {
                withAnnotationScope(LINK_ANNOTATION, href) {
                    withStyleScope(
                        SpanStyle(
                            color = renderer.linkColor,
                            textDecoration = TextDecoration.Underline,
                        ),
                    ) {
                        for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
                    }
                }
            } else {
                for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
            }
        }
        is WireNode.Image -> {
            val alt = node.alt
            withStyleScope(SpanStyle(color = renderer.placeholderColor)) {
                append(if (alt.isEmpty()) "[image]" else "[$alt]")
            }
        }
        WireNode.SoftBreak -> append(' ')
        WireNode.HardBreak -> append('\n')
        is WireNode.Paragraph -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.Heading -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.BlockQuote -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.ListNode,
        is WireNode.CodeBlock,
        WireNode.Rule,
        is WireNode.Media,
        is WireNode.Placeholder -> { /* block-level — never inside an inline reduce */ }
    }
}

/** Resolve a tap offset against the annotations attached during inline build. */
private fun dispatchInlineTap(
    annotated: AnnotatedString,
    offset: Int,
    renderer: NostrContentRenderer,
) {
    annotated.getStringAnnotations(MENTION_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onMentionTap(it.item)
        return
    }
    annotated.getStringAnnotations(EVENT_REF_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onEventRefTap(it.item)
        return
    }
    annotated.getStringAnnotations(HASHTAG_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onHashtagTap(it.item)
        return
    }
    annotated.getStringAnnotations(LINK_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onLinkTap(it.item)
    }
}

// Annotation tags used to round-trip per-run identifiers from
// AnnotatedString into the tap-offset dispatcher. Centralized so the
// builder and the dispatcher cannot drift.
private const val MENTION_ANNOTATION = "nmp:mention"
private const val EVENT_REF_ANNOTATION = "nmp:event_ref"
private const val HASHTAG_ANNOTATION = "nmp:hashtag"
private const val LINK_ANNOTATION = "nmp:link"

/** Push a [SpanStyle] for the duration of [block] then pop it. Avoids the
 *  `androidx.compose.ui.text.withStyle` import to keep this file's API
 *  surface explicit. */
private inline fun AnnotatedString.Builder.withStyleScope(
    style: SpanStyle,
    block: AnnotatedString.Builder.() -> Unit,
) {
    val index = pushStyle(style)
    try {
        block()
    } finally {
        pop(index)
    }
}

/** Push a string annotation for the duration of [block] then pop it. */
private inline fun AnnotatedString.Builder.withAnnotationScope(
    tag: String,
    annotation: String,
    block: AnnotatedString.Builder.() -> Unit,
) {
    val index = pushStringAnnotation(tag, annotation)
    try {
        block()
    } finally {
        pop(index)
    }
}

private fun headingStyle(level: UByte): TextStyle {
    return when (level.toInt()) {
        1 -> TextStyle(fontSize = 30.sp, fontWeight = FontWeight.Bold)
        2 -> TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold)
        3 -> TextStyle(fontSize = 22.sp, fontWeight = FontWeight.SemiBold)
        4 -> TextStyle(fontSize = 19.sp, fontWeight = FontWeight.SemiBold)
        5 -> TextStyle(fontSize = 17.sp, fontWeight = FontWeight.SemiBold)
        else -> TextStyle(fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}

private fun shortEntity(value: String): String {
    if (value.length <= 12) return value
    return "${value.take(8)}…${value.takeLast(4)}"
}
